crates/rspack_plugin_real_content_hash/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+#![feature(let_chains)]
+
+use std::{
+  borrow::Cow,
+  hash::{BuildHasherDefault, Hash, Hasher},
+};
+
+use derivative::Derivative;
+use once_cell::sync::{Lazy, OnceCell};
+use rayon::prelude::*;
+use regex::{Captures, Regex};
+use rspack_core::{
+  rspack_sources::{BoxSource, RawSource, SourceExt},
+  AssetInfo, Plugin, PluginContext, PluginProcessAssetsOutput, ProcessAssetsArgs,
+};
+use rustc_hash::{FxHashMap as HashMap, FxHashSet as HashSet, FxHasher};
+use xxhash_rust::xxh3::Xxh3;
+
+type IndexSet<T> = indexmap::IndexSet<T, BuildHasherDefault<FxHasher>>;
+
+pub static QUOTE_META: Lazy<Regex> =
+  Lazy::new(|| Regex::new(r"[-\[\]\\/{}()*+?.^$|]").expect("Invalid regex"));
+
+#[derive(Debug)]
+pub struct RealContentHashPlugin;
+
+#[async_trait::async_trait]
+impl Plugin for RealContentHashPlugin {
+  async fn process_assets_stage_optimize_hash(
+    &mut self,
+    ctx: PluginContext,
+    args: ProcessAssetsArgs<'_>,
+  ) -> PluginProcessAssetsOutput {
+    self.inner_impl(ctx, args).await
+  }
+}
+
+impl RealContentHashPlugin {
+  async fn inner_impl(
+    &mut self,
+    _ctx: PluginContext,
+    args: ProcessAssetsArgs<'_>,
+  ) -> PluginProcessAssetsOutput {
+    let compilation = args.compilation;
+    let mut hash_to_asset_names: HashMap<&str, Vec<&str>> = HashMap::default();
+    for (name, asset) in compilation
+      .assets()
+      .iter()
+      .filter(|(_, asset)| asset.get_source().is_some())
+    {
+      if let Some(content_hash) = &asset.info.content_hash {
+        hash_to_asset_names
+          .entry(content_hash)
+          .and_modify(|names| names.push(name))
+          .or_insert(vec![name]);
+      }
+    }
+    if hash_to_asset_names.is_empty() {
+      return Ok(());
+    }
+    let hash_regexp = Regex::new(
+      &hash_to_asset_names
+        .keys()
+        .map(|hash| QUOTE_META.replace_all(hash, "\\$0"))
+        .collect::<Vec<Cow<str>>>()
+        .join("|"),
+    )
+    .expect("Invalid regex");
+
+    let assets_data: HashMap<&str, AssetData> = compilation
+      .assets()
+      .par_iter()
+      .filter_map(|(name, asset)| {
+        asset
+          .get_source()
+          .map(|source| (name, source, asset.get_info()))
+      })
+      .map(|(name, source, info)| {
+        (
+          name.as_str(),
+          AssetData::new(source.clone(), info, &hash_regexp),
+        )
+      })
```

Comment:
Looks like this can be merged to the `filter_map` to avoid another unnecessary iteration